4936 Handle bad alerts from cl_send_recap_alerts gracefully

albertisfu opened the pull request:
This PR fixes: #4936 

Based on the data retrieved from the server, the query date retrieved from Redis was `2024-11-15 00:00:00`.

The issue here was that the `alert_sweep:query_date` key was not being cleaned up (it's normally cleared at the end of the whole process). This Redis key was introduced in case the command crashed, allowing it to resume from the exact same date if we wanted to run the command late each day, e.g., at 23:30.

While checking the logs in Loki, I found the root of the problem: there is an alert that currently breaks the `query_alerts` method. This alert is also related to this issue: https://github.com/freelawproject/courtlistener/issues/4698 (I'll add the context I found to that issue).

If there is a bad alert, either due to an unsupported query or bad syntax, the raised Elasticsearch error prevents the command from continuing, which stops alerts from being sent and prevents the cleanup of Redis keys.

To solve this problem, I applied two tweaks:

- Catch the `ApiError` from Elasticsearch that occurs with these kinds of bad queries. This way, the alert is just logged and the system continues processing the remaining alerts.
- In case there’s another unhandled error that hasn’t been encountered yet, I added a 12-hour expiration to the `alert_sweep:query_date` key so it's automatically cleaned up the next time the command runs.
- Finally, I'm considering that `alert_sweep:query_date` might no longer be necessary and could be removed, if we agree to continue querying alerts early each day. In case the command crashes, we’d still have a 24-hour window to investigate the issue and rerun the command. What do you think?

### Querying alerts for the previous day

When checking the logs, I noticed that the command is being run early each day. However, the indexing process and query date were originally designed under the assumption that the command would be run late each day. This mismatch can cause issues similar to the one described in https://github.com/freelawproject/courtlistener/issues/5124.

So I’ve gone ahead and adjusted the command, assuming it will now be run early each day. The re-indexing process and query date are now aligned to use data from the previous day.

mlissner commented:
Code looks pretty good. About this: 

> Finally, I'm considering that alert_sweep:query_date might no longer be necessary and could be removed, if we agree to continue querying alerts early each day. In case the command crashes, we’d still have a 24-hour window to investigate the issue and rerun the command. What do you think?

What's the tradeoff? If we remove it and we don't catch a failure within 24 hours, then what? And if we keep it, what — is the code just more complex? 

The rest sounds great, thank you!

albertisfu commented:
> What's the tradeoff? If we remove it and we don't catch a failure within 24 hours, then what? And if we keep it, what — is the code just more complex?

Yeah, the original idea was that the re-indexing process could take several hours, and if it failed due to an Elasticsearch issue, the command could be restarted manually and resumed. For instance:

Command started on March 28 at 23:00 (sending alerts for the same day)

Re-indexing process failed on March 29 at 01:20

We noticed the failure and ran it again on March 29 at 09:00. Since the date March 28 was stored in Redis, there was no need to tweak the command to run it for the previous day.

But now, since we'll be running the command early in the day for the previous day:

Command starts on March 29 at 00:30 (sending alerts for the previous day, March 28)

Re-indexing process fails on March 29 at 02:20

If we notice the error later, we can restart it anytime on the 29th, and it will still process the previous day (March 28) without needing any tweaks. This is the reason if sending alerts for the previous day the date stored in Redis won’t be needed anymore.

In case we don’t catch the failure within the 24-hour window and we still want to send the missed alerts, we’ll need to tweak the query date and re-run the command. In this case, it might be better to pass the query date as an argument to the command so we can control which day to run it for.

And yeah, removing the Redis key would simplify the code.

Let me know what you think.

mlissner commented:
Sounds good. Let's do it.

albertisfu commented:
Sure! I've removed `alert_sweep:query_date` from the command.

I also introduced the `--query-date` option, so it's now possible to run the command for a specific day other than the previous day if needed:

`cl_send_recap_alerts --query-date  2025-03-28`


We can now clean up the unused Redis key too:
```
from cl.lib.redis_utils import get_redis_interface
r = get_redis_interface("CACHE")
r.delete("alert_sweep:query_date")
```

ERosendo commented on cl/alerts/management/commands/cl_send_recap_alerts.py:
```diff
@@ -699,15 +724,33 @@ def add_arguments(self, parser):
         super().handle(*args, **options)
         testing_mode = options.get("testing_mode", False)
         r = get_redis_interface("CACHE")
+        query_date: datetime.datetime | None = options.get("query_date", None)
```
According to the documentation, the [default](https://docs.python.org/3/library/argparse.html#default) argument of the [add_argument](https://docs.python.org/3/library/argparse.html#the-add-argument-method) method assumes a value of `None` when not explicitly specified,  so we don't need to add any extra default handling.

```suggestion
        query_date: datetime.datetime | None = options["query_date"]
```

albertisfu replied:
Thanks, I've applied the changes.

albertisfu commented:
Thanks @ERosendo! I've applied your suggestion. I also noticed that the test related to sending alerts for a custom day was failing. The issue was due to the timing when the tests run,  it would pass if the local time and UTC fell on the same day, but fail when UTC had already moved to the next day while the local time was still on the previous day.

To fix this, I used `time_machine` to control the dates and applied changes across the tests in `tests_recap_alerts` to make them more reliable.

I also fixed an unrelated flaky test I noticed, which was related to the order of alerts in a subject.

ERosendo reviewed (approved):
LGTM :shipit:
